### PR TITLE
More auto-deploy changes.

### PR DESCRIPTION
**Pull request by evansiroky**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

A few more changes so that auto-deployment works somewhat smoothly in practice. And a refactor to consolidate a bunch of job-related logic into its own util file.

- Move all job-related fields and methods into new JobUtils class.
- Introduce a new `active` field into MonitorableJob to correctly identify active jobs even when jobs are recurring or not yet started.
- Change the way active jobs are found to use that new `active` field.
- When queueing a new DeployJob, wait to create a new one so it doesn't show up in list of jobs for a split second.
- Add new field `autoDeployWithCriticalErrors` to Project and update AutoDeployJob to use this field accordingly
- Add a bunch of new completion messages to the AutoDeployJob
- Only update feed versions in a deployment when the AutoDeployJob starts a DeployJob
- Avoid overwriting deployment data in DeployJob by fetching needed data at point of update

### CODE REVIEW

**Review by br648 — APPROVED** [review]
This looks fine to me.

File: src/main/java/com/conveyal/datatools/manager/models/Project.java [hunk 1]
```diff
@@ -72,6 +72,11 @@
      */
     public Set<AutoDeployType> autoDeployTypes = new HashSet<>();
 
+    /**
+     * Whether to auto-deploy feeds that have critical errors.
+     */
+    public boolean autoDeployWithCriticalErrors = false;
```
Comment (landonreed) [inline src/main/java/com/conveyal/datatools/manager/models/Project.java:78]: Is this a field that we are going to add into the UI or is it just for testing?

**Review by landonreed — APPROVED** [review]
Just a couple of minor comments, but this all seems to make sense!

**Review by binh-dam-ibigroup — APPROVED** [review]
LGTM with some minor comments.

**Comment by binh-dam-ibigroup** [comment]
@br648 I'm assigning this back to you because this PR is against #361 that you created. You can coordinate any changes needed with @evansiroky prior to merging.

File: src/main/java/com/conveyal/datatools/manager/models/Project.java (hunk 1 above)
Comment (evansiroky) [reply src/main/java/com/conveyal/datatools/manager/models/Project.java:78]: Yes, see https://github.com/ibi-group/datatools-ui/pull/637

File: src/main/java/com/conveyal/datatools/manager/jobs/AutoDeployJob.java [hunk 2]
```diff
@@ -163,13 +174,14 @@ public void jobLogic() {
...
             if (shouldWaitForNewFeedVersions) {
+                status.completeSuccessfully("Auto-Deployment will wait for new feed versions to be created from jobs in-progress");
                 return;
             }
 
             // Skip auto-deployment for this project if any of the feed versions contained critical errors.
             if (latestVersionsWithCriticalErrors.size() > 0) {
                 StringBuilder errorMessageBuilder = new StringBuilder(
-                    String.format("Auto deployment skipped for project %s! %s feed(s) contain critical errors:",
+                    String.format("Auto deployment for project %s! %s feed(s) contain critical errors:",
```
[the remark this answers is not in the document]
Comment (evansiroky) [reply src/main/java/com/conveyal/datatools/manager/jobs/AutoDeployJob.java]: The deployment is now not always skipped if there are feeds with critical errors if the `autoDeployWithCriticalErrors` is set to true. Incomplete sentence fixed in c4d04c7f9aeb26c5caa4977f535fac1178ea5684

**Pull request merged**